File: 2nd/homework_1/id_45/111_minimum_depth_of_binary_tree.cpp
/**
 * * Definition for a binary tree node.
 * * struct TreeNode {
 * *     int val;
 * *     TreeNode *left;
 * *     TreeNode *right;
 * *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * * };
 * */
//Attention when one of child depth is 0;
class Solution {
public:
    int minDepth(TreeNode* root) {
        if (!root)
            return 0;
        int left_depth = minDepth(root->left);
        int right_depth = minDepth(root->right);

        if (left_depth > 0 && right_depth > 0)
            return 1 + min(left_depth, right_depth);
        else
            return 1 + (left_depth != 0? left_depth : right_depth);
    }
};

File: 2nd/homework_1/id_69/111_MinimumDepthOfBinaryTree.py
# author: lu19kend
# https://leetcode.com/problems/minimum-depth-of-binary-tree/discuss/


class Solution(object):

    def minDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root is None:
            return 0
        queue = []
        queue.append(root)
        last_width = 1
        cur_depth = 1
        while queue:
            while last_width:
                node = queue.pop(0)
                if node.left is None and node.right is None:
                    return cur_depth
                if node.left is not None:
                    queue.append(node.left)
                if node.right is not None:
                    queue.append(node.right)
                last_width -= 1
            cur_depth += 1
            last_width = len(queue)
